File: projects/chess-coach/apps/harness/src/main/kotlin/harness/StockfishEngine.kt
// SPEC: _spec/chess-coach/harness/components.puml
// SPEC: _spec/chess-coach/harness/lifecycle.puml
package harness

/**
 * Stockfish-specific UCI wrapper for evaluation, legality checking, and FEN derivation.
 */
class StockfishEngine(
    stockfishPath: String = "stockfish",
    private val defaultTimeoutMs: Long = 15_000L,
    private val syzygyPath: String? = System.getenv("SYZYGY_PATH")
) {
    private val driver = UciDriver(listOf(stockfishPath), defaultTimeoutMs)

    fun start() {
        driver.start()
        if (!syzygyPath.isNullOrBlank()) {
            driver.send("setoption name SyzygyPath value $syzygyPath")
        }
    }
    fun stop() = driver.stop()
    val isRunning: Boolean get() = driver.isRunning

    data class EvalResult(val bestMove: String, val cp: Int, val isMate: Boolean, val mateIn: Int = 0)

    fun getEvaluation(fen: String, depth: Int = 15): EvalResult {
        driver.send("position fen $fen")
        driver.send("go depth $depth")

        var bestMove = ""
        var cp = 0
        var isMate = false
        var mateIn = 0

        val lines = driver.readUntil("bestmove", defaultTimeoutMs)
        for (line in lines) {
            if (line.contains("score cp")) {
                try {
                    cp = line.substringAfter("score cp ").substringBefore(" ").toInt()
                } catch (_: Exception) {
                }
            } else if (line.contains("score mate")) {
                try {
                    isMate = true
                    mateIn = line.substringAfter("score mate ").substringBefore(" ").toInt()
                } catch (_: Exception) {
                }
            }
            if (line.startsWith("bestmove")) {
                bestMove = line.substringAfter("bestmove ").substringBefore(" ")
            }
        }
        return EvalResult(bestMove, cp, isMate, mateIn)
    }

    fun checkLegality(fen: String, move: String): Boolean {
        driver.send("position fen $fen moves $move")
        driver.send("go depth 1")

        var isLegal = true
        val lines = driver.readUntil("bestmove", defaultTimeoutMs)
        for (line in lines) {
            if (line.contains("Illegal move")) {
                isLegal = false
            }
        }
        return isLegal
    }

    fun getFenAfterMove(fen: String, uciMove: String): String? {
        driver.send("position fen $fen moves $uciMove")
        driver.send("d")

        var newFen = ""
        val lines = driver.readUntil("Checkers:", defaultTimeoutMs)
        for (line in lines) {
            if (line.startsWith("Fen: ")) {
                newFen = line.substringAfter("Fen: ").trim()
            }
        }

        if (newFen == fen || newFen.isEmpty()) return null
        return newFen
    }
}

File: projects/chess-coach/apps/server/src/StockfishEngine.ts
import { UciDriver } from "./UciDriver.ts";

export type EvalResult = { bestMove: string; cp: number; isMate: boolean; mateIn: number };

/**
 * Stockfish UCI wrapper for evaluation, legality, and FEN derivation.
 * Mirrors the Kotlin StockfishEngine (_spec/api/behavior.md §7.2).
 */
export class StockfishEngine {
  private readonly driver: UciDriver;
  private readonly syzygyPath?: string;

  constructor(
    stockfishPath = "stockfish",
    private readonly defaultTimeoutMs = 15_000,
  ) {
    this.driver = new UciDriver([stockfishPath], defaultTimeoutMs);
    this.syzygyPath = process.env.SYZYGY_PATH || undefined;
  }

  async start(): Promise<void> {
    await this.driver.start();
    if (this.syzygyPath) {
      this.driver.send(`setoption name SyzygyPath value ${this.syzygyPath}`);
    }
  }

  stop(): void {
    this.driver.stop();
  }

  get isRunning(): boolean {
    return this.driver.isRunning;
  }

  async getEvaluation(fen: string, depth = 15): Promise<EvalResult> {
    this.driver.send(`position fen ${fen}`);
    this.driver.send(`go depth ${depth}`);

    let bestMove = "";
    let cp = 0;
    let isMate = false;
    let mateIn = 0;

    const lines = await this.driver.readUntil("bestmove", this.defaultTimeoutMs);
    for (const line of lines) {
      if (line.includes("score cp")) {
        const m = line.match(/score cp (-?\d+)/);
        if (m) cp = parseInt(m[1], 10);
      } else if (line.includes("score mate")) {
        isMate = true;
        const m = line.match(/score mate (-?\d+)/);
        if (m) mateIn = parseInt(m[1], 10);
      }
      if (line.startsWith("bestmove")) {
        bestMove = line.split(" ")[1] ?? "";
      }
    }
    return { bestMove, cp, isMate, mateIn };
  }

  async checkLegality(fen: string, move: string): Promise<boolean> {
    this.driver.send(`position fen ${fen} moves ${move}`);
    this.driver.send("go depth 1");

    let isLegal = true;
    const lines = await this.driver.readUntil("bestmove", this.defaultTimeoutMs);
    for (const line of lines) {
      if (line.includes("Illegal move")) isLegal = false;
    }
    return isLegal;
  }

  async getFenAfterMove(fen: string, uciMove: string): Promise<string | null> {
    this.driver.send(`position fen ${fen} moves ${uciMove}`);
    this.driver.send("d");

    let newFen = "";
    const lines = await this.driver.readUntil("Checkers:", this.defaultTimeoutMs);
    for (const line of lines) {
      if (line.startsWith("Fen: ")) newFen = line.slice(5).trim();
    }

    if (newFen === fen || newFen === "") return null;
    return newFen;
  }
}
